Xorshift whitenoise

Swiftb0y opened the pull request:
inspired by https://youtu.be/Tof5pRedskI?t=2735

The sound is only marginally changed. I also included another `RampingValue` to silence the audible clicks when enabling/disabling the effect (also present `main`).

Swiftb0y committed:
feat(effects): whitenoise: use xorshift instead of std::uniform_real_distribution

Swiftb0y committed:
refactor(effects): whitenoise: improve constness

Swiftb0y committed:
fix(effects): whitenoise: apply ramping dry/wet properly

Swiftb0y committed:
fix(effects): whitenoise: fix clicking sound when en-/disabling the effect

It will now gently fade in/out the noise over the duration of one
buffer.

daschuer commented on src/effects/backends/builtin/whitenoiseeffect.cpp:
```diff
@@ -50,24 +58,40 @@ void WhiteNoiseEffect::processChannel(
     WhiteNoiseGroupState& gs = *pState;
 
-    CSAMPLE drywet = static_cast<CSAMPLE>(m_pDryWetParameter->value());
+    const CSAMPLE drywet = static_cast<CSAMPLE>(m_pDryWetParameter->value());
```
Is this also a gain value? 
```suggestion
    const CSAMPLE_GAIN drywet = static_cast<CSAMPLE_GAIN>(m_pDryWetParameter->value());
```

Swiftb0y replied:
Well, `CSAMPLE_GAIN` is just a typedef alias for `CSAMPLE`... I don't really know what the purpose of the `CSAMPLE` type is at all.

daschuer replied:
The types are explained here: https://github.com/mixxxdj/mixxx/blob/main/src/util/types.h
Does this solve your questions, or do we need to improve the description?

Swiftb0y replied:
Thanks. I only have one question and a couple remarks:
What does the "POT" stand for in "POT array indices" (`SINT` description)?
`SAMPLE` (16-bit) seems virtually unused in our codebase, why do we have it in the first place?
`CSAMPLE_GAIN` should probably a new type-kinda thing with compiletime checks (and in debug-builds) to ensure that the range-invariant is upheld. Maybe something similar for `CSAMPLE` as long as we can ensure they stay zero-cost during runtime.

daschuer replied:
Oh that's a typo 
It should be POD https://en.cppreference.com/w/cpp/named_req/PODType

daschuer replied:
> SAMPLE (16-bit) seems virtually unused in our codebase, why do we have it in the first place?

This is legacy. In early versions of Mixxx Soundsource have produced buffers of SAMPLE. We have changed that at one point. 

For now  CSAMPLE and CSAMPLE_GAIN are only for readability purpose, a lot of room for improvements. I guess all these is a topic of refactoring when we introduce n-channel

Swiftb0y replied:
Thanks for the explanations

daschuer commented on src/effects/backends/builtin/whitenoiseeffect.cpp:
```diff
@@ -50,24 +58,40 @@ void WhiteNoiseEffect::processChannel(
+        for (unsigned int sampleInFrameIndex = 0;
+                sampleInFrameIndex < engineParameters.channelCount();
+                sampleInFrameIndex++) {
+            constexpr float normalization_divisor =
```
float can become CSAMPLE

Swiftb0y replied:
From my understanding, `CSAMPLE` is supposed to represent an audio sample and thus be in the range of `[-1;1]`. Neither, `normalization_divisor` nor the numbers I get from the `xorshift` are in that domain, which is why I chose to not make them a `CSAMPLE`.

daschuer replied:
Oh yes in this case, I have probably commented the wrong line.

daschuer commented on src/effects/backends/builtin/whitenoiseeffect.cpp:
```diff
@@ -50,24 +58,40 @@ void WhiteNoiseEffect::processChannel(
+            const float noise = static_cast<float>(gs.random_state) / normalization_divisor;
 
-        pOutput[i] = pInput[i] * (1 - drywet_ramped) + noise * drywet_ramped;
+            const int sampleIndex =
```
We use SINT for index in context of sample buffer. That may allow vectorization.

Swiftb0y replied:
Sure, but the primary reason for this loop not being vectorized is because the loop iterations are interdependent. This is due to the use of `RampingValue::getNext()`. Similar thing for the actual xorshift since the states are interdepent there as well. Both implementations could be changed to be vectorizer friendly, but I didn't want to spend that effort yet.

daschuer replied:
We use SINT independent from vectorization.

Swiftb0y replied:
Makes sense, though why the `ptrdiff_t`, We're always indexing into a collection, why use a signed type?

daschuer replied:
In c++ array index can be negative, and GCC has refused to vectorise with unsigned in some cases. 

https://stackoverflow.com/questions/3174850/what-is-the-correct-type-for-array-indexes-in-c

daschuer reviewed (changes requested):
Thank you. Not yet tested but I left some type related comments.

Swiftb0y committed:
fix(effects): whitenoise: fix noise polarity
previously the noise effect only produced samples in
the range [0, 1] instead or [-1, 1].

Swiftb0y commented:
Do you think it would make sense to extract the xorshift generator into its separate reusable class?

daschuer commented on src/effects/backends/builtin/whitenoiseeffect.cpp:
```diff
@@ -50,24 +58,41 @@ void WhiteNoiseEffect::processChannel(
+                    static_cast<float>(std::numeric_limits<int32_t>::max());
+            gs.random_state = nextState(gs.random_state);
+            const auto bipolar_random = static_cast<int32_t>(gs.random_state);
+            const float noise = static_cast<float>(bipolar_random) / normalization_divisor;
```
This is the final CSAMPLE
```suggestion
            const CSAMPLE noise = static_cast<float>(bipolar_random) / normalization_divisor;
```

daschuer commented:
> Do you think it would make sense to extract the xorshift generator into its separate reusable class?

Yes, If you are willing to spend the extra effort.

Swiftb0y commented:
I think we can merge without and I'll do that if I have spare time.

Swiftb0y committed:
refactor(effects): whitenoise: use more appropriate typedefs

Swiftb0y commented:
Done

daschuer commented:
This is the spectrum of this PR: 
![grafik](https://user-images.githubusercontent.com/1777442/170608234-1af9cf7b-7904-4f7d-be5f-2e5630a7c344.png)

This is the current main: 
![grafik](https://user-images.githubusercontent.com/1777442/170608694-75f5148d-64f9-4526-a8f2-18f9d32a4bbb.png)

Is the difference expected? Is there a bug in the math? 

With headphones I can hear a minimal difference. I guess it depends on the use case which one we prefer, the fast one, or the cryptographic one. I guess if you are trying to equalize your Setup, you need the later. 
If we just need some noise for dither or such, the first is probably sufficient in favor of saved CPU. 


This is the Audacity build in noise for reference: 
![grafik](https://user-images.githubusercontent.com/1777442/170608442-a5b790ae-9c4b-4c3d-b3f7-1fb370e52106.png)

daschuer commented:
I am not able to explain where the gap in the 80 Hz region comes from. I for a test I have move the conversion to the double domain like this is done in audacity and that does not show this effect. 
Maybe this is a kind of alias effect when throwing away some bits going from int32 to the 24 significant bits of float. 
Using double will use the cut bits for dither. There is also the issue that the negative samples can overshoot. 
We use the invert minimum here to avoid this: 
https://github.com/mixxxdj/mixxx/blob/0ad01b6ccc744045fb727eb4e775aa0e52bf4998/src/util/sample.cpp#L370

daschuer commented:
I think after fixing the conversation, it is fine. At least better than the unpredictable external implementation. 
Thank you for taking this up.

Swiftb0y commented:
I'll look into it when I get the time.
